File: dasan_cli/maintenance.py
"""Maintenance: administration, logs, NTP, backup, firmware, and misc read-only pages."""
import urllib.error
import urllib.request

import typer

from . import core

app = typer.Typer(help="Administration, logs, NTP, backup, firmware, and misc maintenance pages")

ADMIN_PAGE = "MaintainancePage-Administration"
LOGS_PAGE = "MaintainancePage-SystemLog"
NTP_PAGE = "MaintainancePage-NTP"
BACKUP_PAGE = "MaintainancePage-BackupRestore"
FIRMWARE_PAGE = "MaintainancePage-FirmwareUpgrade"
AUTO_REBOOT_PAGE = "MaintainancePage-AutoReboot"
PORT_MIRRORING_PAGE = "MaintainancePage-PortMirroring"
SNMP_PAGE = "MaintainancePage-Snmp"
SYSLOG_CONF_PAGE = "MaintainancePage-SyslogConfiguration"


def _raw_get(client, path):
    """GET a non-JSON (plain text / binary) endpoint the /dm/ JSON wrapper can't handle."""
    req = urllib.request.Request(client.base + path, method="GET")
    if client.token:
        req.add_header("Authorization", f"Bearer {client.token}")
    try:
        with urllib.request.urlopen(req, context=core.SSL_CTX, timeout=15) as resp:
            return resp.read(), resp.headers.get("Content-Type", "")
    except urllib.error.HTTPError as e:
        raise core.RouterError(f"HTTP {e.code} on {path}: {e.read()[:200]}")


@app.command()
def administration(ctx: typer.Context):
    """Web account settings: admin ports, idle timeout, login lockout (read-only)."""
    client = ctx.obj
    d = client.get("Administration", page="", namespace="sys")
    core.console.print(f"[bold]HTTP port:[/bold]        {d['WebPort']} (enabled: {d['HttpEnable']})")
    core.console.print(f"[bold]HTTPS port:[/bold]       {d['HttpsPort']}")
    core.console.print(f"[bold]Access from WAN:[/bold]  {d['WebOnWan']}")
    core.console.print(f"[bold]Idle timeout:[/bold]     {d['timeout']}s")
    core.console.print(f"[bold]Max login trials:[/bold] {d['MaxTrial']} (lockout {d['BannedTime']}s)")


@app.command()
def ntp(ctx: typer.Context):
    """NTP server configuration and sync status (read-only)."""
    client = ctx.obj
    d = client.get("TimeServer", page=NTP_PAGE)
    core.console.print(f"[bold]Enabled:[/bold]      {d['enable']}")
    core.console.print(f"[bold]Status:[/bold]       {d['status']}")
    core.console.print(f"[bold]Current time:[/bold] {d['currentLocalTime']} ({d['localTimeZone']})")
    servers = [d.get(f"NTPServer{i}") for i in range(1, 5) if d.get(f"NTPServer{i}")]
    core.console.print(f"[bold]NTP servers:[/bold]  {', '.join(servers)}")
    core.console.print(f"[bold]DST:[/bold]          {d['daylightSavingsUsed']}")


@app.command()
def firmware(ctx: typer.Context):
    """Current firmware/hardware version (read-only; upgrade must be done via the web UI)."""
    client = ctx.obj
    d = client.get("DeviceInfo")
    core.console.print(f"[bold]Manufacturer:[/bold]     {d['manufacturer']}")
    core.console.print(f"[bold]Model:[/bold]            {d['modelName']}")
    core.console.print(f"[bold]Hardware version:[/bold] {d['hardwareVersion']}")
    core.console.print(f"[bold]Software version:[/bold] {d['softwareVersion']}")


@app.command()
def logs(
    ctx: typer.Context,
    lines: int = typer.Option(50, help="number of most recent log lines to show"),
):
    """Show recent system log entries (read-only)."""
    client = ctx.obj
    raw, _ = _raw_get(client, f"/bin/?objs=SyslogDownload&page={LOGS_PAGE}")
    text = raw.decode("utf-8", errors="replace")
    all_lines = [l for l in text.splitlines() if l.strip()]
    for line in all_lines[-lines:]:
        core.console.print(line)


app.command("system-log")(logs)


@app.command()
def backup(
    ctx: typer.Context,
    output: str = typer.Option(None, help="output file path (default: backup_<timestamp>.bin)"),
):
    """Download the current router configuration backup (GET only, no restore)."""
    client = ctx.obj
    raw, content_type = _raw_get(client, f"/bin/?objs=BackupConfig&page={BACKUP_PAGE}")
    import time
    path = output or f"backup_{time.strftime('%Y%m%d_%H%M%S')}.bin"
    with open(path, "wb") as f:
        f.write(raw)
    core.console.print(f"[green]Saved config backup to {path} ({len(raw)} bytes, {content_type})[/green]")


@app.command("auto-reboot")
def auto_reboot(ctx: typer.Context):
    """Scheduled auto-reboot configuration (read-only)."""
    client = ctx.obj
    d = client.get("AutoRebootObj", page=AUTO_REBOOT_PAGE)
    core.console.print(f"[bold]Enabled:[/bold]        {d['Enable']}")
    core.console.print(f"[bold]Daily at:[/bold]       {d['Hour']:02d}:{d['Minute']:02d}")
    core.console.print(f"[bold]Min uptime:[/bold]     {d['Uptime']}min")
    core.console.print(f"[bold]Uptime-count reboot:[/bold] {d['Enable_ucount']} (threshold {d['Ucount_threshold']})")


@app.command("port-mirroring")
def port_mirroring(ctx: typer.Context):
    """LAN port mirroring configuration (read-only)."""
    client = ctx.obj
    d = client.get("PortMirroring", page=PORT_MIRRORING_PAGE)
    core.console.print(f"[bold]Enabled:[/bold] {d['Enable']}")
    ports = [k.replace("Port_", "") for k in d if k.startswith("Port_") and d[k]]
    core.console.print(f"[bold]Mirrored ports:[/bold] {', '.join(ports) or '-'}")


@app.command()
def snmp(ctx: typer.Context):
    """SNMP agent configuration (read-only; community/password values are masked)."""
    client = ctx.obj
    d = client.get("SnmpCfg", page=SNMP_PAGE)

    def mask(v):
        return "***" if v else "-"

    core.console.print(f"[bold]SNMP v1/v2 enabled:[/bold] {d['snmpActive']}")
    core.console.print(f"[bold]Get community:[/bold]     {mask(d.get('getCommunity'))}")
    core.console.print(f"[bold]Set community:[/bold]     {mask(d.get('setCommunity'))}")
    core.console.print(f"[bold]Trap manager:[/bold]      {d.get('trapManagerIPv4')}")
    core.console.print(f"[bold]sysName/Contact/Loc:[/bold] {d.get('sysName')} / {d.get('sysContact')} / {d.get('sysLocation')}")
    core.console.print(f"[bold]SNMPv3 enabled:[/bold]     {d['snmpV3Active']}")
    if d["snmpV3Active"]:
        core.console.print(f"[bold]SNMPv3 user:[/bold]        {d.get('snmpV3UserName')}")
        core.console.print(f"[bold]SNMPv3 password:[/bold]    {mask(d.get('snmpV3Password'))}")


@app.command("syslog-configuration")
def syslog_configuration(ctx: typer.Context):
    """Remote syslog forwarding configuration (read-only; password is masked)."""
    client = ctx.obj
    d = client.get("RemoteSyslogCfg", page=SYSLOG_CONF_PAGE)
    core.console.print(f"[bold]Enabled:[/bold]  {d['isActive']}")
    core.console.print(f"[bold]Protocol:[/bold] {d['protocol']}")
    core.console.print(f"[bold]Host:[/bold]     {d['host'] or '-'}")
    core.console.print(f"[bold]Level:[/bold]    {d['level']}")
    if d.get("userName"):
        core.console.print(f"[bold]User:[/bold]     {d['userName']} (password: {'***' if d.get('passwd') else '-'})")

File: internal/cli/maintenance.go
package cli

import (
	"fmt"
	"os"
	"time"

	"github.com/spf13/cobra"

	"github.com/anshuman852/dasan/internal/client"
)

// NewMaintenanceCmd creates the `dasan maintenance` command tree.
func NewMaintenanceCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "maintenance",
		Short: "Administration, logs, NTP, backup, firmware, and misc maintenance pages",
	}

	cmd.AddCommand(
		newMaintAdminCmd(),
		newMaintNtpCmd(),
		newMaintFirmwareCmd(),
		newMaintLogsCmd(),
		newMaintBackupCmd(),
		newMaintAutoRebootCmd(),
		newMaintPortMirroringCmd(),
		newMaintSnmpCmd(),
		newMaintSyslogConfigCmd(),
	)

	return cmd
}

func newMaintAdminCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "admin",
		Short: "Web account settings: admin ports, idle timeout, login lockout (read-only)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			d, err := cl.GetNS("Administration", "", "sys")
			if err != nil {
				return err
			}
			obj, err := client.ToObj(d)
			if err != nil {
				return err
			}
			fmt.Printf("HTTP port:        %s (enabled: %v)\n", client.GetStr(obj, "WebPort"), obj["HttpEnable"])
			fmt.Printf("HTTPS port:       %s\n", client.GetStr(obj, "HttpsPort"))
			fmt.Printf("Access from WAN:  %v\n", obj["WebOnWan"])
			fmt.Printf("Idle timeout:     %s s\n", client.GetStr(obj, "timeout"))
			fmt.Printf("Max login trials: %s (lockout %s s)\n", client.GetStr(obj, "MaxTrial"), client.GetStr(obj, "BannedTime"))
			return nil
		},
	}
}

func newMaintNtpCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "ntp",
		Short: "NTP server configuration and sync status (read-only)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("TimeServer", "MaintainancePage-NTP")
			if err != nil {
				return err
			}
			d, err := client.ToObj(data)
			if err != nil {
				return err
			}
			fmt.Printf("Enabled:      %v\n", d["enable"])
			fmt.Printf("Status:       %s\n", client.GetStr(d, "status"))
			fmt.Printf("Current time: %s (%s)\n", client.GetStr(d, "currentLocalTime"), client.GetStr(d, "localTimeZone"))
			servers := []string{}
			for i := 1; i <= 4; i++ {
				key := fmt.Sprintf("NTPServer%d", i)
				if v, ok := d[key]; ok && fmt.Sprintf("%v", v) != "" {
					servers = append(servers, fmt.Sprintf("%v", v))
				}
			}
			if len(servers) == 0 {
				servers = append(servers, "-")
			}
			fmt.Printf("NTP servers:  %s\n", joinStrings(servers, ", "))
			fmt.Printf("DST:          %v\n", d["daylightSavingsUsed"])
			return nil
		},
	}
}

func newMaintFirmwareCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "firmware",
		Short: "Current firmware/hardware version (read-only)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("DeviceInfo", "")
			if err != nil {
				return err
			}
			d, err := client.ToObj(data)
			if err != nil {
				return err
			}
			fmt.Printf("Manufacturer:     %s\n", client.GetStr(d, "manufacturer"))
			fmt.Printf("Model:            %s\n", client.GetStr(d, "modelName"))
			fmt.Printf("Hardware version: %s\n", client.GetStr(d, "hardwareVersion"))
			fmt.Printf("Software version: %s\n", client.GetStr(d, "softwareVersion"))
			return nil
		},
	}
}

func newMaintLogsCmd() *cobra.Command {
	var lines int
	cmd := &cobra.Command{
		Use:   "logs",
		Short: "Show recent system log entries (read-only)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			raw, err := cl.GetRaw("/bin/?objs=SyslogDownload&page=MaintainancePage-SystemLog")
			if err != nil {
				return err
			}
			text := string(raw)
			// Print last N non-empty lines
			allLines := []string{}
			for _, l := range textLines(text) {
				if l != "" {
					allLines = append(allLines, l)
				}
			}
			start := len(allLines) - lines
			if start < 0 {
				start = 0
			}
			for _, l := range allLines[start:] {
				fmt.Println(l)
			}
			return nil
		},
	}
	cmd.Flags().IntVar(&lines, "lines", 50, "number of most recent log lines to show")
	return cmd
}

func newMaintBackupCmd() *cobra.Command {
	var output string
	cmd := &cobra.Command{
		Use:   "backup",
		Short: "Download the current router configuration backup (GET only, no restore)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			raw, err := cl.GetRaw("/bin/?objs=BackupConfig&page=MaintainancePage-BackupRestore")
			if err != nil {
				return err
			}
			path := output
			if path == "" {
				path = fmt.Sprintf("backup_%s.bin", time.Now().Format("20060102_150405"))
			}
			if err := os.WriteFile(path, raw, 0644); err != nil {
				return fmt.Errorf("write backup: %w", err)
			}
			fmt.Printf("Saved config backup to %s (%d bytes)\n", path, len(raw))
			return nil
		},
	}
	cmd.Flags().StringVar(&output, "output", "", "output file path")
	return cmd
}

func newMaintAutoRebootCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "auto-reboot",
		Short: "Scheduled auto-reboot configuration (read-only)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("AutoRebootObj", "MaintainancePage-AutoReboot")
			if err != nil {
				return err
			}
			d, err := client.ToObj(data)
			if err != nil {
				return err
			}
			fmt.Printf("Enabled:              %v\n", d["Enable"])
			fmt.Printf("Daily at:             %02d:%02d\n", int(client.GetFloat(d, "Hour")), int(client.GetFloat(d, "Minute")))
			fmt.Printf("Min uptime:           %.0f min\n", client.GetFloat(d, "Uptime"))
			fmt.Printf("Uptime-count reboot:  %v (threshold %.0f)\n", d["Enable_ucount"], client.GetFloat(d, "Ucount_threshold"))
			return nil
		},
	}
}

func newMaintPortMirroringCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "port-mirroring",
		Short: "LAN port mirroring configuration (read-only)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("PortMirroring", "MaintainancePage-PortMirroring")
			if err != nil {
				return err
			}
			d, err := client.ToObj(data)
			if err != nil {
				return err
			}
			fmt.Printf("Enabled: %v\n", d["Enable"])
			ports := []string{}
			for k, v := range d {
				if len(k) > 5 && k[:5] == "Port_" && isTruthy(v) {
					ports = append(ports, k[5:])
				}
			}
			if len(ports) == 0 {
				fmt.Println("Mirrored ports: -")
			} else {
				fmt.Printf("Mirrored ports: %s\n", joinStrings(ports, ", "))
			}
			return nil
		},
	}
}

func newMaintSnmpCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "snmp",
		Short: "SNMP agent configuration (read-only; values are masked)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("SnmpCfg", "MaintainancePage-Snmp")
			if err != nil {
				return err
			}
			d, err := client.ToObj(data)
			if err != nil {
				return err
			}
			mask := func(v string) string {
				if v != "" {
					return "***"
				}
				return "-"
			}
			fmt.Printf("SNMP v1/v2 enabled: %v\n", d["snmpActive"])
			fmt.Printf("Get community:      %s\n", mask(client.GetStr(d, "getCommunity")))
			fmt.Printf("Set community:      %s\n", mask(client.GetStr(d, "setCommunity")))
			fmt.Printf("Trap manager:       %s\n", client.GetStr(d, "trapManagerIPv4"))
			fmt.Printf("sysName/Contact/Loc: %s / %s / %s\n", client.GetStr(d, "sysName"), client.GetStr(d, "sysContact"), client.GetStr(d, "sysLocation"))
			fmt.Printf("SNMPv3 enabled:     %v\n", d["snmpV3Active"])
			if client.GetBool(d, "snmpV3Active") {
				fmt.Printf("SNMPv3 user:        %s\n", client.GetStr(d, "snmpV3UserName"))
				fmt.Printf("SNMPv3 password:    %s\n", mask(client.GetStr(d, "snmpV3Password")))
			}
			return nil
		},
	}
}

func newMaintSyslogConfigCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "syslog-configuration",
		Short: "Remote syslog forwarding configuration (read-only; password is masked)",
		RunE: func(cmd *cobra.Command, args []string) error {
			cl := getClient(cmd)
			data, err := cl.Get("RemoteSyslogCfg", "MaintainancePage-SyslogConfiguration")
			if err != nil {
				return err
			}
			d, err := client.ToObj(data)
			if err != nil {
				return err
			}
			fmt.Printf("Enabled:  %v\n", d["isActive"])
			fmt.Printf("Protocol: %s\n", client.GetStr(d, "protocol"))
			host := client.GetStr(d, "host")
			if host == "" {
				host = "-"
			}
			fmt.Printf("Host:     %s\n", host)
			fmt.Printf("Level:    %s\n", client.GetStr(d, "level"))
			if client.GetStr(d, "userName") != "" {
				pw := "-"
				if client.GetStr(d, "passwd") != "" {
					pw = "***"
				}
				fmt.Printf("User:     %s (password: %s)\n", client.GetStr(d, "userName"), pw)
			}
			return nil
		},
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func textLines(s string) []string {
	lines := []string{}
	current := ""
	for _, c := range s {
		if c == '\n' {
			lines = append(lines, current)
			current = ""
		} else {
			current += string(c)
		}
	}
	if current != "" {
		lines = append(lines, current)
	}
	return lines
}

func isTruthy(v any) bool {
	switch t := v.(type) {
	case bool:
		return t
	case string:
		return t == "true" || t == "True" || t == "1" || t == "yes"
	case float64:
		return t != 0
	default:
		return false
	}
}

func joinStrings(items []string, sep string) string {
	if len(items) == 0 {
		return "-"
	}
	result := ""
	for i, s := range items {
		if i > 0 {
			result += sep
		}
		result += s
	}
	return result
}
